### PR TITLE
WT-11285 Introduce WT_READ_ONCE macro

### PR DESCRIPTION
**Pull request by ajmorton**
`WT_ORDERED_READ` is currently being used for two purposes in the code base: 
1. To ensure that read operation A takes place before read operation B, preventing the compiler and CPU from reordering the two reads for optimization purposes, and
2. Ensuring that a single read operation in the code becomes a single read operation in the compiled output. Disallowing the compiler from splitting one read into multiple, or fusing two reads into one (volatile semantics).

This PR adds the new macro `WT_READ_ONCE` to handle case 1). Over time this will replace uses of `WT_ORDERED_READ` read that only provide volatile semantics, and `WT_ORDERED_READ` will only be used to handle case 2).

### CODE REVIEW

File: src/conn/conn_log.c [hunk 1]
```diff
@@ -404,7 +404,7 @@ __log_remove_once(WT_SESSION_IMPL *session, uint32_t backup_file)
     WT_ORDERED_READ(dbg_val, conn->debug_ckpt_cnt);
     if (FLD_ISSET(conn->debug_flags, WT_CONN_DEBUG_CKPT_RETAIN) && dbg_val != 0)
         min_lognum = WT_MIN(conn->debug_ckpt[dbg_val - 1].l.file, min_lognum);
-    WT_ORDERED_READ(dbg_val, conn->debug_log_cnt);
+    WT_READ_ONCE(dbg_val, conn->debug_log_cnt);
```
Comment (luke-pearson) [inline src/conn/conn_log.c:407]: I still think we should do this in your debug_blah ticket.

File: src/include/hardware.h [hunk 2]
```diff
@@ -16,6 +16,32 @@
         (v) = (val);        \
     } while (0)
 
+/*
+ * WT_READ_ONCE --
+ *
+ * Ensure a single read from memory in the source code produces a single read from memory in the
+ * compiled output.
+ *
+ * The compiler is allowed to optimize loads from memory in multiple ways such as Load fusing, where
```
Comment (luke-pearson) [inline src/include/hardware.h:25]: I think using quotes here to designate naming a thing would be better. So such as "load fusing" and "invented loads".

Comment (ajmorton) [reply src/include/hardware.h]: Added in [c460ce8](https://github.com/wiredtiger/wiredtiger/pull/9413/commits/c460ce81236e86dbce5ca0e35af8610fbd0a4260)

File: src/conn/conn_log.c (hunk 1 above)
Comment (ajmorton) [reply src/conn/conn_log.c:407]: After offline discussion we'll remove this line and move the PR to draft mode. We can merge the PR when a future ticket has need for this macro

File: src/include/hardware.h [hunk 3]
```diff
@@ -16,6 +16,32 @@
...
+ *
+ * GCC and Clang have a __typeof__ compiler builtin which allows us to temporarily cast the value
+ * being read as a volatile and achieve volatile semantics. For other compilers we'll fall back on
+ * inserted a read barrier after the read (our pre-existing implementation) which prevents invented
+ * and fused loads for this variable in the code following the expression.
+ */
+#if defined(__GNUC__) || defined(__clang__)
+#define WT_READ_ONCE(v, val) (v) = (*(volatile __typeof__(val) *)&(val));
+#else
+#define WT_READ_ONCE(v, val) WT_ORDERED_READ(v, val)
```
Comment (luke-pearson) [inline src/include/hardware.h:42]: I was thinking about this but don't see it as a natural path forward without causing perf issues on windows. For example WT-11414 wants to replace volatile generations with read_once. But if we did that everywhere the code would slow down on windows? So we couldn't introduce it there. There's also another PR #9456 that did have a volatile which could've used READ_ONCE, it went with locking instead but I don't know how comfortable I'd be introducing a tonne of read barriers on Windows.

**Comment by ajmorton** [comment]
After discussion we'll move ahead with the current implementation, and once our Windows platforms that support C11 generics are available we'll update WT_READ_ONCE to have a common implementation using _Generics on all platforms. 

Since this is an interim fix, and most if not all of our planned changes will be updating existing `WT_ORDERED_READ`s to `WT_READ_ONCE` there won't be many new barrier added to the code. We also expect platforms to be ready well before the next release at which point we'll be using the correct volatile implementation.

File: src/include/hardware.h (hunk 3 above)
Comment (luke-pearson) [reply src/include/hardware.h:48]: As a result of being unable to implement READ_ONCE on windows due to some MSVC issues we're going ahead with a read barrier.

File: src/include/hardware.h [hunk 4]
```diff
@@ -17,7 +17,67 @@
...
+ * By placing a read barrier we ensure that any reads before the barrier take place before any
+ * reads after the barrier, and as a result can guarantee that this specific read happens before any
+ * reads that follow this instruction.
+ * Note that this makes no guarantees about:
+ *   - Ordering between this read and earlier reads in the source code.
+ *   - Ordering between this read subsequent non-read instructions, for example writes to memory.
+ *
+ * FIXME-WT-10946 - Add a link to the shared memory documentation once complete.
+ */
+#define WT_READ_ORDERED(v, val) \
```
Comment (luke-pearson) [inline src/include/hardware.h:71]: Why is this not just replacing ORDERED_READ, is that for demonstration reasons?

Comment (ajmorton) [reply src/include/hardware.h]: As discussed we'll push `WT_READ_ORDERED` out into a separate ticket. I've removed it from this PR.

**Review by quchenhao — APPROVED** [review]
lgtm

**Review by luke-pearson — APPROVED** [review]
LGTM.

**Pull request merged**